### DISCUSSION TITLE
How does `solveFor()` work?

### DISCUSSION THREAD

**Discussion by ChocolateLoverRaj** [jiggzson/nerdamer#663] [Q&A] [answered]
I'm curious to know how `nerdamer('some equation or expression').solveFor('some variable')` works. Especially for a really complicated equation that has a really complicated solution like `((initialX1 + t * velecityX1) - (initialX2 + t * velocityX2))^2 + ((initialY1 + t * velocityY1) - (initialY2 + t * velocityY2))^2 = (radius1 + radius2) ^ 2`. I looked at the source code https://github.com/jiggzson/nerdamer/blob/81a7b14c36d98741a8d49ebf73a63adf8bb6f612/Solve.js but the file is super big and I can't understand from looking at the file.

**Comment by jiggzson** [comment, marked as answer]
>I looked at the source code https://github.com/jiggzson/nerdamer/blob/81a7b14c36d98741a8d49ebf73a63adf8bb6f612/Solve.js but the file is super big and I can't understand from looking at the file.

I agree. The library should be refactored, however is quite a time intensive undertaking. There is an PR (#643) where @mugabe has made quite a bit of progress. Unfortunately it's only the core at the moment. 

>I'm curious to know how nerdamer('some equation or expression').solveFor('some variable') works.

If you mean the flow for `solveFor` then it's simply an extension of the `Expression` class [here](https://github.com/jiggzson/nerdamer/blob/0c1b6e54725592de0e4b81a5a2639d0abb4e5201/Solve.js#L238) and after doing some factoring, rewriting, etc. it eventually simply applies the quadratic equation [here](https://github.com/jiggzson/nerdamer/blob/0c1b6e54725592de0e4b81a5a2639d0abb4e5201/Solve.js#L1623). The actual function is [here](https://github.com/jiggzson/nerdamer/blob/0c1b6e54725592de0e4b81a5a2639d0abb4e5201/Solve.js#L792)

I'm going to assume that you mean how `nerdamer` solves the expression/equation in the hope you can extract only the parts that you need. Honestly, solve is just a composition of a bunch of solving algorithms. The quadratic, cubic, quartic, Jenkins-Traubb (I know there's some duplication of the other methods here which I hope to change some day), Newton's method, etc. The biggest reason why just extracting the parts that you need might prove difficult is because of all the prep work before the solving algorithms are applied. One of the approaches is to simply factor the expression and then solve the factors, so you need `factor`. Factoring is proving to be quite challenging by the way at least for the multivariate kind. In fact the strength of the library heavily depends on `factor`. Factoring requires polynomial division, again not trivial. And so forth.

In short, it can be done but it will require a bit of work and I'd love to provide insight should you decide to do so.

**Reply by ChocolateLoverRaj** [reply]
> quadratic, cubic, quartic, Jenkins-Traubb

So the equation in my example is solved using one of these methods?

**Comment by jiggzson** [comment]
>>quadratic, cubic, quartic, Jenkins-Traubb
>
>So the equation in my example is solved using one of these methods?

Correct. The quadratic equation in your case but after it's been rewritten. So it goes in as `((a+x*v1)-(b+x*v2))^2+((c+x*v3)-(d+x*v4))^2-(r1+r2)^2` but ultimately gets transformed/expanded to `-2*a*b-2*a*v2*x-2*b*v1*x-2*c*d-2*c*v4*x-2*d*v3*x-2*r1*r2-2*v1*v2*x^2-2*v3*v4*x^2-r1^2-r2^2+2*a*v1*x+2*b*v2*x+2*c*v3*x+2*d*v4*x+a^2+b^2+c^2+d^2+v1^2*x^2+v2^2*x^2+v3^2*x^2+v4^2*x^2`. The coefficients are calculated and the equation is then applied.

**Reply by ChocolateLoverRaj** [reply]
Wow. I thought that equation was too long to be just a quadratic!
